### ISSUE TITLE
Possible memory leak when writing a lot of data

### ISSUE THREAD

**Issue by MarvinJWendt** [pocketbase/pocketbase#969]
Hi, we tested our production app with a load tester (https://k6.io). It seems like memory isn't fully freed after many (many) writes, which results in constant memory increase and a possible DoS.

### K6 load test scenario

We were testing the API endpoint `/api/collections/REDACTED/records`, with an `expand` and `sort` statement. 

#### Reproducible K6 script

```js
import http from 'k6/http';
import { sleep } from 'k6';

export const options = {
  stages: [
    { duration: '10s', target: 1400 }, // spike to 1400 users
    { duration: '3m', target: 1400 }, // stay at 1400 for 3 minutes
    { duration: '10s', target: 0 }, // scale down. Recovery stage.
  ],
};
export default function () {
  const BASE_URL = 'https://REDACTED.com';

  const responses = http.batch([
    ['GET', `${BASE_URL}/api/collections/REDACTED/records?expand=REDACTED&sort=-REDACTED`, null, { tags: { name: 'api' } }],
  ]);

  sleep(1);
}
```

#### Memory usage of PocketBase

![image](https://user-images.githubusercontent.com/31022056/200666050-cec61784-4a8c-4dc1-a41c-6d09286f7ef0.png)

// EDIT: I just noticed that the screenshot cut off the time, so here it is:
![image](https://user-images.githubusercontent.com/31022056/200667544-10430981-a010-4830-a194-4f18216a3c1d.png)

#### Database size after stress testing
![image](https://user-images.githubusercontent.com/31022056/200664717-577d4999-89be-44b7-8697-3e7f9215d7f0.png)

#### Database size after forced server restart
![image](https://user-images.githubusercontent.com/31022056/200666276-7b713bdb-f0e5-44b7-954a-8f5bc8392b77.png)

## Note

As the server restart cleared the memory, and it did not increase again, I suppose something isn't freed right, while PocketBase is running.

### Build info

PocketBase was build without CGO in a docker environment (golang:alpine).

**Comment by ganigeorgiev** [comment]
What version of PocketBase are you using? In the recent v0.7.10 and v0.8.0-rc2 there were several improvements to keep the memory and open file descriptors low. If you are not using any of those version, please retest again.

Also idle memory increase may be just virual memory that the os/gc keeps. We can force the garbage collector to release the memory with `debug.FreeOSMemory` but I don't think it's necessary and it's better to leave the runtime to decide when to release the memory.

**Comment by MarvinJWendt** [comment]
> What version of PocketBase are you using?

Sorry, forgot to mention it, I am using `v0.7.10`.

> Also idle memory increase may be just "virtual memory" that the os/gc keeps.

I also thought about that, but the memory increase took over 30 minutes and was about 5 GB, then we restarted the container. Also, the `.wal` file, had about 100M the whole 30 minutes, until the restart happened. I am not an SQLite expert, so I am just noting this down, in case it helps.

**Comment by MarvinJWendt** [comment]
>  We "soft" limited journal size to ~100mb, but in practice it can grow indefinitely on heave concurrent use.

Yes, while the stress test was running, we had a temporary 20 GB increase of disk usage by PocketBase.

> You could try disabling the logs write by going to "Admin UI > Settings > Logs max days retention" to 0 and retest again.

I did that now. I don't see any logs anymore, but the `logs.db-wal` sticks to 566M. Shouldn't this just be wiped when the log retention is set to 0?

I restarted the server to provide a clean test result for the load test, after clearing the logs. `logs.db` is still 595M, but the UI doesn't show any logs.

![image](https://user-images.githubusercontent.com/31022056/200677525-ea3cc649-2985-4b84-a4e0-c5fccd2ed665.png)


The load test (with disabled logs) just increased the memory barely (<0.3%).

**Comment by ganigeorgiev** [comment]
> I did that now. I don't see any logs anymore, but the logs.db-wal sticks to 566M. Shouldn't this just be wiped when the log retention is set to 0?

The `db-wal` and `db-shm` are managed by the sqlite process and eventually will be deleted when all connections to the database are closed (aka. on application shutdown/restart). 

> I restarted the server to provide a clean test result for the load test, after clearing the logs. logs.db is still 595M, but the UI doesn't show any logs.

Hm, that's strange. It should be deleted, maybe there is some meta data related to the internal `_requests` table that sqlite is keeping to reduce future writes, I'm not really sure.

Tomorrow after work I'll install k6 locally and will test it to see if we can improve something.

**Comment by ganigeorgiev** [comment]
About why the `logs.db` size is not reduced after deleting the rows - it is because auto_vacuum is disabled by default and sqlite is keeping the space for subsequent writes. From https://sqlite.org/pragma.html#pragma_auto_vacuum:

> When auto-vacuum is disabled and data is deleted data from a database, **the database file remains the same size**. Unused database file pages are added to a "[freelist](https://sqlite.org/fileformat2.html#freelist)" and **reused for subsequent inserts**. So no database file space is lost. However, the database file does not shrink. In this mode the [VACUUM](https://sqlite.org/lang_vacuum.html) command can be used to rebuild the entire database file and thus reclaim unused disk space.

_I'll experiment with this tomorrow to see if it will increase the loading times with large db files, but eventually we may try running `VACUUM` at least for the `logs.db` when the admin updates the logs retention._

**Comment by MarvinJWendt** [comment]
How about also adding a `VACUUM` button in the settings (not just for the logs)? That way, admins could trigger it manually, when a lot of data has been removed. This could be useful when backups are created, to save space.

Uptime Kuma has this option too: 
![image](https://user-images.githubusercontent.com/31022056/200686936-d7b867b8-205d-486a-99d9-6f83e356fd7e.png)

**Comment by ganigeorgiev** [comment]
**TLDR:** I couldn't find memory leaks in the PocketBase source. Inspecting the allocation with the pprof tool and runtime.MemStats also didn't show anything too unusual that will indicate a memory leak.
When running the stress test multiple times, without restarting the server, also showed **no significant memory accumulation** over time and it pretty much stayed on the same idle level after each test.
Additionally, **running `debug.FreeOSMemory()` in a separate go routine periodically almost immediately releases the idle memory**, which should be an indication that the released memory is not actually used and it is just some runtime optimization technique to limit eventual subsequent allocations.

----

First some info on my test setup:

- I've increased the ulimit to 5000 (aka. `ulimit -n 5000`) because the default on my system is 1024 and in order to run 1400 concurrent requests you'll need at least the same amount of file descriptors

- I have created 2 test collections (`posts` and `authors`) each with 15k records and single relation field in posts->author (the total db size is ~5mb).

- I'm using the `rc` branch (aka. v0.8.0), but for the targeted GET request the app behavior should be more or less the same compared to v0.7.10

- I'm running k6 on the same machine in another terminal by `k6 run script.js`, where `script.js` is similar from your example:

  ```js
  import { sleep } from 'k6';
  import http from 'k6/http';

  export const options = {
    setupTimeout: '120s',
    stages: [
      { duration: '10s', target: 1400 },
      { duration: '60s', target: 1400 },
      { duration: '5s', target: 0 },
    ],
  };

  export default function () {
    http.batch([
      ['GET', `http://localhost:8090/api/collections/posts/records?perPage=200&expand=author&sort=-created`, null, { tags: { name: 'api' } }],
    ]);

    sleep(1);
  }
  ```

1. From just looking at htop, the memory consumption growth is up to ~1.5-2GB while the stress test is running (**please keep in mind that this is while htop and k6 is running on the same machine; htop indicates that the k6 process takes ~15-20% of the total memory usage**). During the tests, the memory consumption starts from ~2GB -> reaches ~3.8-4GB -> then after completion stays at ~2,8-3GB. Debugging with `pprof` and `runtime.MemStats` also shows similar proportional memory growth when looking at the heap and allocs data.

2. To confirm that this idle memory (in my case as shown above is ~1gb) is not actually used memory space, I've added in a separate go routine `debug.FreeOSMemory()` to run periodically. Then I rerun the test and the idle memory got released almost immediately after the `FreeOSMemory` call (htop showed drop from ~3gb -> ~2.2-2.3gb total system memory consumption).

3. To further confirm that the idle memory doesn't accumulate over time, I've executed the stress test 3 more times, without restarting the PocketBase process. The result after each test were similar - starts from ~2GB reaching up to ~4GB total system memory consumption, then back to ~3GB after the test and after `debug.FreeOSMemory()` - back to ~2-2.2GB.

While there is definitely room for improvement and allocations could be further optimized, I don't think there is a memory leak (or at least I cannot identify it).

Debugging the heap and allocs with pprof showed that the json serialization of `models.Record` requires surprisingly large number of memory allocations. Additionally `DateTime.Scan` could be also improved because when populating the timestamp fields from the raw DB data it seems to reduce in too much memory pressure. These and other minor things will be improved in the future but that will be after the migration flow and filters enhancements.

---

About the SQLite `VACUUM` - I want to avoid users to have manually take care for that and adding a manual trigger would be only as a last resort. I've tested `auto_vacuum=full` but it's not the magic solution that I expected it to be and it can actually made the performance slightly worst because it increased the exclusive lock times (based on the official docs it could also worsen the data fragmentation but I didn't notice any significant query performance degradation related to that). `auto_vacuum` has "incremental" option that will batch the required operations and eventually could reduce the lock times. It needs to be combined with the `incremental_vacuum(NPages)` pragma but I haven't really tested it and I don't know what would be the optimal number of pages in our case. I've added in the [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1) as a reminder to investigate it more thoroughly and eventually, as a last resort, adding a "Maintenance" section for performing some system actions like manually running vacuum, db reindexing, etc. (it could be also eventually combined with the backup feature request from #338).

For now, I'll add `VACUUM` to run in the background only if the logs retention option has been changed to "0" (even if the `VACUUM` operation takes some time to complete it should be safe because it is guaranteed that there would be no logs.db write during the execution and shouldn't interrupt the usual application flow).

--- 

As a footnote, if you are concerned about the performance:
- Try building with `CGO_ENABLED=1` (the cgo sqlite driver is ~2-3 times faster then the noncgo version; it complicates the cross-compilation, but if you are already building and running in a container it may not be an issue for your)
- Go 1.19 comes with `debug.SetMemoryLimit(...)` which could eventually reduce the max idle memory kept by the runtime

**Comment by samducker** [comment]
Hi @ganigeorgiev I found a similar issue today whilst doing some load testing. 

I was firing off some load testing and the performance held up well,

Here is one of my tests getting a joined table (users and posts)

<img width="954" alt="CleanShot 2023-10-02 at 04 02 53@2x" src="https://github.com/pocketbase/pocketbase/assets/25123915/2cbdf4b8-c7a4-4806-9b41-a0ca8dfb5e6a">

So I made a lot of requests in a short space of time. Which naturally spiked the CPU and the memory, however the bad news is the memory stayed peaked for hours after and would only come down after a restart of the server.

<img width="1053" alt="CleanShot 2023-10-02 at 03 58 53@2x" src="https://github.com/pocketbase/pocketbase/assets/25123915/79d2a0e1-45aa-4327-bbbc-1ff532e5364e">

Any ideas on why this happened and how it could clear the memory better on its own, so as not to use additional resource.

**Comment by ganigeorgiev** [comment]
We don't have direct control over the memory and when the Go runtime will release it (see also https://github.com/golang/go/issues/42805).

To force the GC to run more often you can enable the [`GOMEMLIMIT=500MiB` env variable](_https://pkg.go.dev/runtime) (or fine tune its behavior with GOGCC but I'd recommend against it if you are not sure what you are doing_).

**Comment by samducker** [comment]
Thanks @ganigeorgiev 

The analysis above makes sense however, I was surprised that the GC had not run literally hours after the test.

Is this likely that it would have eventually cleared by itself or would it have always required a restart?

**Comment by ganigeorgiev** [comment]
GC cycles are not necessary "time based". The Go GC scheduler will not release the memory if it doesn't need to (eg. when the underlying OS has plenty of memory left to work with).

If you are concerned about OOM errors consider enabling the `GOMEMLIMIT` env variable. It is a soft-limit and will instruct the GC to try to run whenever the specified limit is reached.
